File: dotnet/cluster_management/examples/CreateSingleRegionCluster/CreateSingleRegionCluster.cs
using System.Diagnostics;
using Amazon;
using Amazon.DSQL;
using Amazon.DSQL.Model;
using Amazon.Runtime.Credentials;

namespace DSQLExamples.CreateSingleRegionCluster;

public class CreateSingleRegionCluster
{
    /// <summary>
    /// Create a client. We will use this later for performing operations on the cluster.
    /// </summary>
    private static async Task<AmazonDSQLClient> CreateDSQLClient(RegionEndpoint region)
    {
        var awsCredentials = await DefaultAWSCredentialsIdentityResolver.GetCredentialsAsync();
        var clientConfig = new AmazonDSQLConfig
        {
            RegionEndpoint = region
        };
        return new AmazonDSQLClient(awsCredentials, clientConfig);
    }

    /// <summary>
    /// Create an example DSQL cluster.
    /// </summary>
    public static async Task<CreateClusterResponse> Create(RegionEndpoint region)
    {
        using var client = await CreateDSQLClient(region);

        var repo = Environment.GetEnvironmentVariable("GITHUB_REPOSITORY") ?? "local";
        var runId = Environment.GetEnvironmentVariable("GITHUB_RUN_ID") ?? "local";

        var createClusterRequest = new CreateClusterRequest
        {
            DeletionProtectionEnabled = true,
            Tags = new Dictionary<string, string>
            {
                { "Name", "csharp single region cluster" },
                { "Repo", repo },
                { "Type", "cluster-management" },
                { "RunId", runId }
            }
        };

        CreateClusterResponse response = await client.CreateClusterAsync(createClusterRequest);
        Console.WriteLine($"Initiated creation of {response.Arn}");

        return response;
    }

    public static async Task Main()
    {
        var regionName = Environment.GetEnvironmentVariable("CLUSTER_REGION");
        Debug.Assert(!string.IsNullOrEmpty(regionName), "Environment variable `CLUSTER_REGION` must be set");
        var region = RegionEndpoint.GetBySystemName(regionName);

        await Create(region);
    }
}

File: rust/cluster_management/src/bin/create_single_region_cluster.rs
use aws_config::load_defaults;
use aws_sdk_dsql::client::Waiters;
use aws_sdk_dsql::operation::get_cluster::GetClusterOutput;
use aws_sdk_dsql::{
    Client, Config,
    config::{BehaviorVersion, Region},
};
use std::collections::HashMap;
use std::env;

/// Create a client. We will use this later for performing operations on the cluster.
async fn dsql_client(region: &str) -> anyhow::Result<Client> {
    // Load default SDK configuration
    let sdk_defaults = load_defaults(BehaviorVersion::latest()).await;

    // You can set your own credentials by following this guide
    // https://docs.aws.amazon.com/sdk-for-rust/latest/dg/credproviders.html
    let credentials = sdk_defaults
        .credentials_provider()
        .ok_or_else(|| anyhow::anyhow!("Failed to obtain credentials provider"))?;

    let config = Config::builder()
        .behavior_version(BehaviorVersion::latest())
        .credentials_provider(credentials)
        .region(Region::new(region.to_owned()))
        .build();

    Ok(Client::from_conf(config))
}

/// Create an example DSQL cluster.
pub async fn create_cluster(region: &str) -> anyhow::Result<GetClusterOutput> {
    let client = dsql_client(region).await?;
    let repo = env::var("GITHUB_REPOSITORY").unwrap_or_else(|_| "local".to_string());
    let run_id = env::var("GITHUB_RUN_ID").unwrap_or_else(|_| "local".to_string());
    let tags = HashMap::from([
        (
            String::from("Name"),
            String::from("rust single region cluster"),
        ),
        (String::from("Repo"), repo),
        (String::from("Type"), String::from("cluster-management")),
        (String::from("RunId"), run_id),
    ]);

    let create_cluster_output = client
        .create_cluster()
        .set_tags(Some(tags))
        .deletion_protection_enabled(true)
        .send()
        .await?;
    println!("Created {}", create_cluster_output.arn);

    println!("Waiting for cluster to become ACTIVE");
    let cluster = client
        .wait_until_cluster_active()
        .identifier(&create_cluster_output.identifier)
        .wait(std::time::Duration::from_secs(300)) // Wait up to 5 minutes
        .await?
        .into_result()?;

    Ok(cluster)
}

#[tokio::main(flavor = "current_thread")]
pub async fn main() -> anyhow::Result<()> {
    let region = env::var("CLUSTER_REGION").expect("env variable `CLUSTER_REGION` should be set");

    let output = create_cluster(&region).await?;
    println!("{:#?}", output);
    Ok(())
}
